File: 14_lmm_sabr_hybrid_model/cpp/SABRSimulation.cpp
#include "SABRSimulation.hpp"

#include <algorithm>
#include <cmath>
#include <random>
#include <stdexcept>


std::size_t SABRSimulationResult::numberOfPaths() const
{
    return forwardPaths.size();
}


std::size_t SABRSimulationResult::numberOfTimes() const
{
    if (forwardPaths.empty())
    {
        return 0;
    }

    return forwardPaths.front().size();
}


std::size_t SABRSimulationResult::numberOfForwards() const
{
    if (forwardPaths.empty() ||
        forwardPaths.front().empty())
    {
        return 0;
    }

    return forwardPaths
        .front()
        .front()
        .size();
}


Vector buildTimeGrid(
    double simulationEnd,
    std::size_t numberOfSteps,
    const Vector& mandatoryTimes
)
{
    if (simulationEnd <= 0.0)
    {
        throw std::invalid_argument(
            "simulationEnd must be positive."
        );
    }

    if (numberOfSteps == 0)
    {
        throw std::invalid_argument(
            "numberOfSteps must be positive."
        );
    }

    Vector times;

    times.reserve(
        numberOfSteps
        +
        mandatoryTimes.size()
        +
        1
    );

    double dt =
        simulationEnd
        /
        static_cast<double>(
            numberOfSteps
        );

    for (std::size_t i = 0;
         i <= numberOfSteps;
         ++i)
    {
        times.push_back(
            static_cast<double>(i)
            *
            dt
        );
    }

    for (double mandatoryTime :
         mandatoryTimes)
    {
        if (mandatoryTime >= 0.0 &&
            mandatoryTime <= simulationEnd)
        {
            times.push_back(
                mandatoryTime
            );
        }
    }

    std::sort(
        times.begin(),
        times.end()
    );

    Vector uniqueTimes;

    for (double time :
         times)
    {
        if (uniqueTimes.empty() ||
            std::fabs(
                time
                -
                uniqueTimes.back()
            )
            >
            1.0e-12)
        {
            uniqueTimes.push_back(
                time
            );
        }
    }

    return uniqueTimes;
}


std::size_t exactTimeIndex(
    const Vector& times,
    double targetTime,
    double tolerance
)
{
    for (std::size_t i = 0;
         i < times.size();
         ++i)
    {
        if (std::fabs(
                times[i]
                -
                targetTime
            )
            <=
            tolerance)
        {
            return i;
        }
    }

    throw std::runtime_error(
        "The requested time is not "
        "present on the simulation grid."
    );
}


SABRSimulationResult simulateLMMSABR(
    const LMMSABRHybrid& model,
    double simulationEnd,
    std::size_t numberOfSteps,
    std::size_t numberOfPaths,
    unsigned int seed,
    bool antithetic,
    const Vector& mandatoryTimes
)
{
    if (numberOfPaths == 0)
    {
        throw std::invalid_argument(
            "numberOfPaths must be positive."
        );
    }

    Vector times =
        buildTimeGrid(
            simulationEnd,
            numberOfSteps,
            mandatoryTimes
        );

    std::size_t numberOfTimes =
        times.size();

    std::size_t numberOfForwards =
        model.numberOfForwards();

    std::size_t numberOfFactors =
        model.numberOfFactors();

    SABRSimulationResult result;

    result.times = times;

    result.forwardPaths =
        Tensor3D(
            numberOfPaths,
            std::vector<Vector>(
                numberOfTimes,
                Vector(
                    numberOfForwards,
                    0.0
                )
            )
        );

    result.alphaPaths =
        Tensor3D(
            numberOfPaths,
            std::vector<Vector>(
                numberOfTimes,
                Vector(
                    numberOfForwards,
                    0.0
                )
            )
        );

    for (std::size_t path = 0;
         path < numberOfPaths;
         ++path)
    {
        result.forwardPaths[path][0] =
            model.initialForwards();

        result.alphaPaths[path][0] =
            model.sabrParameters().alpha0();
    }

    std::mt19937 generator(
        seed
    );

    std::normal_distribution<double>
        standardNormal(
            0.0,
            1.0
        );

    for (std::size_t timeIndex = 0;
         timeIndex + 1 < numberOfTimes;
         ++timeIndex)
    {
        double currentTime =
            times[timeIndex];

        double dt =
            times[timeIndex + 1]
            -
            times[timeIndex];

        Vector storedRateShocks;
        Vector storedVolatilityShocks;

        for (std::size_t path = 0;
             path < numberOfPaths;
             ++path)
        {
            Vector rateFactorShocks(
                numberOfFactors,
                0.0
            );

            Vector independentVolatilityShocks(
                numberOfForwards,
                0.0
            );

            bool useAntitheticPath =
                antithetic
                &&
                path % 2 == 1;

            if (!useAntitheticPath)
            {
                for (std::size_t factor = 0;
                     factor < numberOfFactors;
                     ++factor)
                {
                    rateFactorShocks[factor] =
                        standardNormal(
                            generator
                        );
                }

                for (std::size_t i = 0;
                     i < numberOfForwards;
                     ++i)
                {
                    independentVolatilityShocks[i] =
                        standardNormal(
                            generator
                        );
                }

                storedRateShocks =
                    rateFactorShocks;

                storedVolatilityShocks =
                    independentVolatilityShocks;
            }
            else
            {
                for (std::size_t factor = 0;
                     factor < numberOfFactors;
                     ++factor)
                {
                    rateFactorShocks[factor] =
                        -storedRateShocks[factor];
                }

                for (std::size_t i = 0;
                     i < numberOfForwards;
                     ++i)
                {
                    independentVolatilityShocks[i] =
                        -storedVolatilityShocks[i];
                }
            }

            Vector nextForwards;
            Vector nextAlpha;

            model.evolve(
                currentTime,
                result.forwardPaths
                    [path]
                    [timeIndex],
                result.alphaPaths
                    [path]
                    [timeIndex],
                dt,
                rateFactorShocks,
                independentVolatilityShocks,
                nextForwards,
                nextAlpha
            );

            result.forwardPaths
                [path]
                [timeIndex + 1]
                =
                nextForwards;

            result.alphaPaths
                [path]
                [timeIndex + 1]
                =
                nextAlpha;
        }
    }

    return result;
}

File: 14_lmm_sabr_hybrid_model/python/sabr_simulation.py
"""
sabr_simulation.py

Monte Carlo simulation for the LMM-SABR hybrid.

The simulator stores:

    forward_paths
        path x time x forward

    alpha_paths
        path x time x forward

The simulation grid can include mandatory dates so that:

    reset dates
    caplet fixing dates
    swaption expiries

are represented exactly.
"""

import numpy as np


class LMMSABRSimulationResult:
    """
    Store LMM-SABR simulation results.
    """

    def __init__(
        self,
        times,
        forward_paths,
        alpha_paths
    ):

        self.times = np.asarray(
            times,
            dtype=float
        )

        self.forward_paths = np.asarray(
            forward_paths,
            dtype=float
        )

        self.alpha_paths = np.asarray(
            alpha_paths,
            dtype=float
        )

        self.validate()

    def validate(
        self
    ):
        """
        Validate simulation dimensions.
        """

        if self.forward_paths.ndim != 3:

            raise ValueError(
                "forward_paths must be three-dimensional."
            )

        if self.alpha_paths.ndim != 3:

            raise ValueError(
                "alpha_paths must be three-dimensional."
            )

        if (
            self.forward_paths.shape
            !=
            self.alpha_paths.shape
        ):

            raise ValueError(
                "Forward and alpha paths "
                "must have equal dimensions."
            )

        if (
            self.forward_paths.shape[1]
            !=
            len(
                self.times
            )
        ):

            raise ValueError(
                "The time dimension is inconsistent."
            )

    def get_number_of_paths(
        self
    ):

        return self.forward_paths.shape[0]

    def get_number_of_times(
        self
    ):

        return self.forward_paths.shape[1]

    def get_number_of_forwards(
        self
    ):

        return self.forward_paths.shape[2]


def build_time_grid(
    simulation_end,
    number_of_steps,
    mandatory_times=None
):
    """
    Build a sorted simulation time grid.
    """

    if simulation_end <= 0.0:

        raise ValueError(
            "simulation_end must be positive."
        )

    if number_of_steps <= 0:

        raise ValueError(
            "number_of_steps must be positive."
        )

    base_grid = np.linspace(
        0.0,
        simulation_end,
        number_of_steps + 1
    )

    if mandatory_times is None:

        return base_grid

    mandatory_times = np.asarray(
        mandatory_times,
        dtype=float
    )

    mandatory_times = mandatory_times[
        (
            mandatory_times >= 0.0
        )
        &
        (
            mandatory_times <= simulation_end
        )
    ]

    combined_grid = np.concatenate(
        [
            base_grid,
            mandatory_times
        ]
    )

    combined_grid = np.unique(
        combined_grid
    )

    return np.sort(
        combined_grid
    )


def simulate_lmm_sabr(
    model,
    simulation_end,
    number_of_steps,
    number_of_paths,
    seed=42,
    antithetic=True,
    mandatory_times=None
):
    """
    Simulate forward rates and stochastic volatilities.
    """

    if number_of_paths <= 0:

        raise ValueError(
            "number_of_paths must be positive."
        )

    if mandatory_times is None:

        mandatory_times = (
            model.get_reset_times()
        )

    times = build_time_grid(
        simulation_end,
        number_of_steps,
        mandatory_times
    )

    number_of_times = len(
        times
    )

    forward_paths = np.empty(
        (
            number_of_paths,
            number_of_times,
            model.number_of_forwards
        )
    )

    alpha_paths = np.empty(
        (
            number_of_paths,
            number_of_times,
            model.number_of_forwards
        )
    )

    forward_paths[:, 0, :] = (
        model.initial_forwards
    )

    alpha_paths[:, 0, :] = (
        model.sabr_parameters.alpha0
    )

    random_generator = (
        np.random.default_rng(
            seed
        )
    )

    half_number_of_paths = (
        number_of_paths + 1
    ) // 2

    for time_index in range(
        number_of_times - 1
    ):

        current_time = float(
            times[time_index]
        )

        dt = float(
            times[time_index + 1]
            -
            times[time_index]
        )

        if antithetic:

            base_rate_shocks = (
                random_generator.normal(
                    size=(
                        half_number_of_paths,
                        model.number_of_factors
                    )
                )
            )

            base_volatility_shocks = (
                random_generator.normal(
                    size=(
                        half_number_of_paths,
                        model.number_of_forwards
                    )
                )
            )

            rate_shocks = np.concatenate(
                [
                    base_rate_shocks,
                    -base_rate_shocks
                ],
                axis=0
            )

            volatility_shocks = np.concatenate(
                [
                    base_volatility_shocks,
                    -base_volatility_shocks
                ],
                axis=0
            )

            rate_shocks = rate_shocks[
                :number_of_paths
            ]

            volatility_shocks = (
                volatility_shocks[
                    :number_of_paths
                ]
            )

        else:

            rate_shocks = (
                random_generator.normal(
                    size=(
                        number_of_paths,
                        model.number_of_factors
                    )
                )
            )

            volatility_shocks = (
                random_generator.normal(
                    size=(
                        number_of_paths,
                        model.number_of_forwards
                    )
                )
            )

        for path_index in range(
            number_of_paths
        ):

            current_forwards = (
                forward_paths[
                    path_index,
                    time_index,
                    :
                ]
            )

            current_alpha = (
                alpha_paths[
                    path_index,
                    time_index,
                    :
                ]
            )

            next_forwards, next_alpha = (
                model.evolve(
                    current_time,
                    current_forwards,
                    current_alpha,
                    dt,
                    rate_shocks[
                        path_index
                    ],
                    volatility_shocks[
                        path_index
                    ]
                )
            )

            forward_paths[
                path_index,
                time_index + 1,
                :
            ] = next_forwards

            alpha_paths[
                path_index,
                time_index + 1,
                :
            ] = next_alpha

    return LMMSABRSimulationResult(
        times,
        forward_paths,
        alpha_paths
    )


def exact_time_index(
    times,
    target_time,
    tolerance=1.0e-10
):
    """
    Find a target time on the simulation grid.
    """

    times = np.asarray(
        times,
        dtype=float
    )

    differences = np.abs(
        times
        -
        float(
            target_time
        )
    )

    index = int(
        np.argmin(
            differences
        )
    )

    if differences[index] > tolerance:

        raise ValueError(
            f"Target time {target_time} "
            "is not on the simulation grid."
        )

    return index